### ISSUE TITLE
ProjectScanner must check if file is directory

### ISSUE THREAD

**Issue by stephanj** [devoxx/DevoxxGenieIDEAPlugin#369]
```
Error reading file: /Users/stephan/IdeaProjects/llama3.java java.io.IOException: Cannot get content of directory: file:///Users/stephan/IdeaProjects/llama3.java at com.intellij.openapi.vfs.newvfs.impl.VirtualDirectoryImpl.contentsToByteArray(VirtualDirectoryImpl.java:599) at com.devoxx.genie.service.projectscanner.ProjectScannerService$1.lambda$visitFile$0(ProjectScannerService.java:228) at
```

**Comment by stephanj** [comment]
Solution was found by Google Gemini 2.0 Flash 🥳 🧠

The error message `java.io.IOException: Cannot get content of directory: file:///Users/stephan/IdeaProjects/llama3.java` indicates that the code is trying to treat a **file** (`llama3.java`) as a **directory**. This is happening because `llama3.java` is likely being passed to `VfsUtilCore.visitChildrenRecursively` which expects a directory.

Here's a breakdown of the problem and the solution:

**Problem:**

The `walkThroughDirectory` method uses `VfsUtilCore.visitChildrenRecursively` to traverse the file system. This method is designed for directories. When it encounters a file, it still tries to visit its "children" (which a file doesn't have), leading to the `IOException`. This likely happens when the starting point for scanning is incorrectly set to a file instead of a directory.

**Solution:**

We need to modify the logic to correctly handle cases where the starting point (`startDirectory`) might be a file instead of a directory.

Here's the updated code with the fix:

```java
import com.devoxx.genie.model.ScanContentResult;
import com.devoxx.genie.service.DevoxxGenieSettingsService;
import com.devoxx.genie.ui.settings.DevoxxGenieStateService;
import com.devoxx.genie.util.GitignoreParser;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.application.ReadAction;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectUtil;
import com.intellij.openapi.roots.ProjectFileIndex;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.vfs.VfsUtilCore;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileVisitor;
import com.intellij.util.concurrency.AppExecutorUtil;
import com.knuddels.jtokkit.Encodings;
import com.knuddels.jtokkit.api.Encoding;
import com.knuddels.jtokkit.api.EncodingType;
import com.knuddels.jtokkit.api.IntArrayList;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;

public class ProjectScannerService {

    private static final Logger LOG = Logger.getInstance(ProjectScannerService.class.getName());

    private static final Encoding ENCODING = Encodings.newDefaultEncodingRegistry().getEncoding(EncodingType.CL100K_BASE);
    private static final String GITIGNORE = ".gitignore";

    private GitignoreParser gitignoreParser;

    public static ProjectScannerService getInstance() {
        return ApplicationManager.getApplication().getService(ProjectScannerService.class);
    }

    public CompletableFuture<ScanContentResult> scanProject(Project project,
                                                            VirtualFile startDirectory,
                                                            int windowContextMaxTokens,
                                                            boolean isTokenCalculation) {
        CompletableFuture<ScanContentResult> future = new CompletableFuture<>();
        ScanContentResult scanContentResult = new ScanContentResult();

        ApplicationManager.getApplication().executeOnPooledThread(() -> {
            try {
                ReadAction.nonBlocking(() -> {
                            StringBuilder result = new StringBuilder();
                            StringBuilder fullContent;

                            initGitignoreParser(project, startDirectory);

                            if (startDirectory == null) {
                                result.append("Directory Structure:\n");
                                fullContent = getContentFromModules(project, windowContextMaxTokens, result, scanContentResult);
                            } else if (startDirectory.isDirectory()) {
                                result.append("Directory Structure:\n");
                                fullContent = processDirectory(project, startDirectory, result, scanContentResult, windowContextMaxTokens);
                            } else {
                                // Handle the case where startDirectory is a file
                                result.append("File:\n");
                                fullContent = processSingleFile(project, startDirectory, result, scanContentResult);
                            }

                            String content = isTokenCalculation ? fullContent.toString() :
                                    truncateToTokens(fullContent.toString(), windowContextMaxTokens, isTokenCalculation);

                            scanContentResult.setTokenCount(ENCODING.countTokens(content));
                            scanContentResult.setContent(content);

                            return scanContentResult;
                        })
                        .inSmartMode(project)
                        .finishOnUiThread(ModalityState.defaultModalityState(), future::complete)
                        .submit(AppExecutorUtil.getAppExecutorService());
            } catch (Exception e) {
                future.completeExceptionally(e);
            }
        });

        return future;
    }

    public ScanContentResult scanProjectSynchronously(Project project, VirtualFile startDirectory, int windowContextMaxTokens, boolean isTokenCalculation) {
        ScanContentResult scanContentResult = new ScanContentResult();
        ReadAction.run(() -> {
            StringBuilder result = new StringBuilder();
            StringBuilder fullContent;
            initGitignoreParser(project, startDirectory);

            if (startDirectory == null) {
                result.append("Directory Structure:\n");
                fullContent = getContentFromModules(project, windowContextMaxTokens, result, scanContentResult);
            } else if (startDirectory.isDirectory()) {
                result.append("Directory Structure:\n");
                fullContent = processDirectory(project, startDirectory, result, scanContentResult, windowContextMaxTokens);
            } else {
                // Handle the case where startDirectory is a file
                result.append("File:\n");
                fullContent = processSingleFile(project, startDirectory, result, scanContentResult);
            }

            String content = isTokenCalculation ? fullContent.toString() : truncateToTokens(fullContent.toString(), windowContextMaxTokens, isTokenCalculation);
            scanContentResult.setTokenCount(ENCODING.countTokens(content));
            scanContentResult.setContent(content);
        });
        return scanContentResult;
    }

    private void initGitignoreParser(Project project, VirtualFile startDirectory) {

        VirtualFile gitignoreFile = null;
        if (startDirectory != null && startDirectory.isDirectory()) {
            gitignoreFile = startDirectory.findChild(GITIGNORE);
        } else if (project != null) {
            gitignoreFile = Objects.requireNonNull(ProjectUtil.guessProjectDir(project)).findChild(GITIGNORE);
        } else {
            LOG.error("Error initializing GitignoreParser: .gitignore file could not be found or startDirectory is not a directory");
        }

        if (gitignoreFile != null && gitignoreFile.exists()) {
            try {
                gitignoreParser = new GitignoreParser(Paths.get(gitignoreFile.getPath()));
            } catch (IOException e) {
                LOG.error("Error initializing GitignoreParser: " + e.getMessage());
            }
        }
    }

    private StringBuilder getContentFromModules(Project project,
                                                int windowContextMaxTokens,
                                                StringBuilder result,
                                                ScanContentResult scanContentResult) {

        UniqueDirectoryScannerService uniqueDirectoryScanner = new UniqueDirectoryScannerService();

        // Collect all content roots from modules
        VirtualFile[] contentRootsFromAllModules =
                ProjectRootManager.getInstance(project).getContentRootsFromAllModules();

        // Add all content roots to the unique directory scanner
        Arrays.stream(contentRootsFromAllModules)
                .distinct()
                .forEach(uniqueDirectoryScanner::addDirectory);

        // Get the highest root directory and process the content
        return uniqueDirectoryScanner
                .getHighestCommonRoot()
                .map(highestCommonRoot -> processDirectory(project,
                        highestCommonRoot,
                        result,
                        scanContentResult,
                        windowContextMaxTokens))
                .orElseThrow();
    }

    private @NotNull StringBuilder processDirectory(Project project,
                                                    VirtualFile startDirectory,
                                                    @NotNull StringBuilder result,
                                                    ScanContentResult scanContentResult,
                                                    int windowContextMaxTokens) {
        result.append(generateSourceTreeRecursive(startDirectory, 0));

        result.append("\n\nFile Contents:\n");

        ProjectFileIndex fileIndex = ProjectFileIndex.getInstance(project);

        StringBuilder fullContent = new StringBuilder(result);
        AtomicInteger currentTokens = new AtomicInteger(0);

        walkThroughDirectory(startDirectory, fileIndex, fullContent, currentTokens, scanContentResult);
        return fullContent;
    }

    private @NotNull StringBuilder processSingleFile(Project project,
                                                      VirtualFile file,
                                                      @NotNull StringBuilder result,
                                                      ScanContentResult scanContentResult) {
        result.append(file.getName()).append("\n");
        result.append("\n\nFile Contents:\n");

        StringBuilder fullContent = new StringBuilder(result);
        if (shouldIncludeFile(file)) {
            readFileContent(file, fullContent, scanContentResult);
        } else {
            scanContentResult.incrementSkippedFileCount();
            LOG.info("Skipping file: " + file.getPath() + " (excluded by settings or .gitignore)");
        }
        return fullContent;
    }

    private void walkThroughDirectory(@NotNull VirtualFile directory,
                                      @NotNull ProjectFileIndex fileIndex,
                                      @NotNull StringBuilder fullContent,
                                      @NotNull AtomicInteger currentTokens,
                                      @NotNull ScanContentResult scanContentResult) {

        VfsUtilCore.visitChildrenRecursively(directory, new VirtualFileVisitor<Void>() {
            @Override
            public boolean visitFile(@NotNull VirtualFile file) {
                if (file.isDirectory()) {
                    if (shouldExcludeDirectory(file)) {
                        scanContentResult.incrementSkippedDirectoryCount();
                        return false;
                    }
                } else {
                    if (fileIndex.isInContent(file) && !shouldExcludeFile(file) && shouldIncludeFile(file)) {
                        readFileContent(file, fullContent, scanContentResult);
                    } else {
                        scanContentResult.incrementSkippedFileCount();
                        LOG.info("Skipping file: " + file.getPath() + " (not in content, excluded by settings, or .gitignore)");
                    }
                }
                return true;
            }
        });
    }

    private void readFileContent(@NotNull VirtualFile file, @NotNull StringBuilder fullContent, @NotNull ScanContentResult scanContentResult) {
        scanContentResult.incrementFileCount();
        scanContentResult.addFile(Paths.get(file.getPath()));

        String header = "\n--- " + file.getPath() + " ---\n";
        fullContent.append(header);

        try {
            String content = ReadAction.compute(() -> {
                try {
                    return new String(file.contentsToByteArray(), StandardCharsets.UTF_8);
                } catch (IOException e) {
                    LOG.error("Error reading file: " + file.getPath(), e);
                    return "Error reading file content: " + e.getMessage();
                }
            });

            content = processFileContent(content);
            fullContent.append(content).append("\n");
        } catch (Exception e) {
            String errorMsg = "Error processing file: " + e.getMessage() + "\n";
            fullContent.append(errorMsg);
        }
    }

    private String truncateToTokens(String text,
                                    int windowContext,
                                    boolean isTokenCalculation) {
        IntArrayList tokens = ENCODING.encode(text);
        if (tokens.size() <= windowContext) {
            return text;
        }

        IntArrayList truncatedTokens = new IntArrayList(windowContext);
        for (int i = 0; i < windowContext; i++) {
            truncatedTokens.add(tokens.get(i));
        }

        String truncatedContent = ENCODING.decode(truncatedTokens);
        return isTokenCalculation ? truncatedContent :
                truncatedContent + "\n--- Project context truncated due to token limit ---\n";
    }

    private @NotNull String generateSourceTreeRecursive(VirtualFile virtualFile, int depth) {
        StringBuilder result = new StringBuilder();
        String indent = "  ".repeat(depth);

        boolean excludeFile = shouldExcludeFile(virtualFile);
        boolean excludeDirectory = shouldExcludeDirectory(virtualFile);
        if (excludeFile || excludeDirectory) {
            return "";
        }

        result.append(indent).append(virtualFile.getName()).append("/\n");

        for (VirtualFile child : virtualFile.getChildren()) {
            if (child.isDirectory()) {
                result.append(generateSourceTreeRecursive(child, depth + 1));
            } else if (shouldIncludeFile(child)) {
                result.append(indent).append("  ").append(child.getName()).append("\n");
            }
        }

        return result.toString();
    }

    private boolean shouldExcludeDirectory(@NotNull VirtualFile file) {
        DevoxxGenieSettingsService settings = DevoxxGenieStateService.getInstance();
        return file.isDirectory() &&
                (settings.getExcludedDirectories().contains(file.getName()) || shouldExcludeFile(file));
    }

    private boolean shouldExcludeFile(@NotNull VirtualFile file) {
        DevoxxGenieSettingsService settings = DevoxxGenieStateService.getInstance();

        // Check if file is in excluded files list
        if (settings.getExcludedFiles().contains(file.getName())) {
            return true;
        }

        // Check gitignore if enabled
        if (Boolean.TRUE.equals(settings.getUseGitIgnore()) &&
                gitignoreParser != null) {
            Path path = Paths.get(file.getPath());
            return gitignoreParser.matches(path);
        }
        return false;
    }

    private boolean shouldIncludeFile(@NotNull VirtualFile file) {
        DevoxxGenieSettingsService settings = DevoxxGenieStateService.getInstance();

        // First check if file should be excluded
        if (shouldExcludeFile(file)) {
            return false;
        }

        // Then check if extension is included
        String extension = file.getExtension();
        return extension != null && settings.getIncludedFileExtensions().contains(extension.toLowerCase());
    }

    private String processFileContent(String content) {
        if (Boolean.TRUE.equals(DevoxxGenieStateService.getInstance().getExcludeJavaDoc())) {
            return removeJavadoc(content);
        }
        return content;
    }

    private @NotNull String removeJavadoc(String content) {
        // Remove block comments (which include Javadoc)
        content = content.replaceAll("/\\*{1,2}[\\s\\S]*?\\*/", "");
        // Remove single-line comments that start with '///'
        content = content.replaceAll("^\\s*///.*$", "");
        return content;
    }
}
```

**Key Changes:**

1. **`scanProject` and `scanProjectSynchronously`:**
    *   Added a check: `else if (startDirectory.isDirectory())`.
    *   Added a new `else` block to handle the case where `startDirectory` is a file, calling `processSingleFile`.

2. **`processSingleFile` Method:**
    *   This new method handles the case where the `startDirectory` is a single file.
    *   It adds the file name to the `result` and then calls `readFileContent` to process the file's content.

3. **`readFileContent` Method:**
    *   This method encapsulates the logic for reading and processing a single file's content.
    *   It includes error handling for `IOException` during file reading.

4. **`walkThroughDirectory` Method:**
    *   Now correctly distinguishes between directories and files.
    *   Calls `readFileContent` only for files that should be included.
    *   Logs skipped files with a reason.

5. **`initGitignoreParser` Method:**
    *   Added a check to ensure that `startDirectory` is a directory before attempting to find the `.gitignore` file within it.

**Explanation:**

*   The main issue was that the code assumed `startDirectory` would always be a directory. The added `isDirectory()` check and the `processSingleFile` method now handle cases where a file is provided as the starting point.
*   `processSingleFile` directly processes the file content without trying to traverse it as a directory.
*   `readFileContent` handles the actual reading and processing of the file's content, making the code more modular.
*   The changes in `walkThroughDirectory` ensure that files are only processed if they meet the inclusion criteria and that skipped files are logged appropriately.
*   The `initGitignoreParser` now handles the case where the start directory might be null or a file, preventing potential errors.

With these changes, the code should now correctly handle both directory and file inputs for `startDirectory` and avoid the `IOException: Cannot get content of directory` error. Remember to thoroughly test the changes with different project structures and file types.

**Comment by stephanj** [comment]
Fix included in v0.4.3
